lib/pdn/pdn.c

```
@@ -0,0 +1,763 @@
+
+		if ((event != PDN_EVENT_NA) && (p_pdn->callback != NULL)) {
+			p_pdn->callback(p_pdn, event);
+		}
+	}
+}
+
+static struct pdn_connection *connection_allocate(void)
+{
+	/* Reserve PDN handle */
+	struct pdn_connection *const p_pdn =
+		malloc(sizeof(struct pdn_connection));
+
+	if (p_pdn == NULL) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	memset(p_pdn, 0, sizeof(struct pdn_connection));
+	p_pdn->state = PDN_ALLOCATED;
+	p_pdn->family = PDN_AF_INET | PDN_AF_INET6;
+
+	/* Set PDN to indicate invalid/unread value. */
+	p_pdn->pdn_id = PDN_ID_UNINITIALIZED;
+
+	/* Register AT event handler */
+	at_notif_register_handler((void *)p_pdn, _at_notif_handler);
+
+	return p_pdn;
+}
+
+static void connection_free(void **handle)
+{
+	if (*handle != NULL) {
+		/* Remove AT event handler */
+		at_notif_deregister_handler(*handle, _at_notif_handler);
+
+		/* Free PDN handle */
+		free(*handle);
+		*handle = NULL;
+	}
+}
+
+static int command_response_parse(struct pdn_connection *const handle,
+				  char *const p_resp,
+				  const enum pdn_management_at_evt evt_id)
+{
+	const char *p_compare_str = m_pdn_mananagement_at_evts[evt_id].p_at_str;
+	const int compare_length = strlen(p_compare_str);
+
+	if (strlen(p_resp) >= compare_length) {
+		if (strncmp(p_compare_str, p_resp, compare_length) == 0) {
+			/* Found the event we are waiting for.
+			 * Parse parameters if any and return.
+			 */
+			return m_pdn_mananagement_at_evts[evt_id].parser_fn(
+				handle, p_resp);
+		}
+	}
+
+	if (strncmp("ERROR", p_resp, strlen("ERROR")) == 0) {
+		errno = ENOEXEC;
+		return -1;
+	}
+
+	errno = EIO;
+	return -1;
+}
+
+static int command_prepare_and_send(struct pdn_connection *handle,
+				    const enum pdn_management_at_cmd cmd_id,
+				    const char *p_remote_addr)
+{
+	int err = 0;
+	int offset = 0;
+	enum pdn_management_at_evt resp_evt = EVT_OK;
+
+	char *const p_command = (char *)malloc(PDN_AT_BUFFER_SIZE);
+
+	if (p_command == NULL) {
+		errno = ENOMEM;
```

Do not manipulate `errno`; we are now in application space we can just return the appropriate error.
Also, errors should be negative.

---

ok, I will fix these.